api/src/main/java/org/apache/iceberg/actions/RewriteManifests.java

```diff
@@ -44,6 +47,43 @@ public interface RewriteManifests
    */
   RewriteManifests rewriteIf(Predicate<ManifestFile> predicate);
 
+  /**
+   * Rewrite manifests in a given order, based on partition field names
+   *
+   * <p>Supply an optional set of partition field names to cluster the rewritten manifests by. For
+   * example, given a table PARTITIONED BY (a, b, c, d), you may wish to rewrite and cluster
+   * manifests by ('d', 'b') only, based on your query patterns. Rewriting Manifests in this way
+   * will yield manifest_lists that point to manifest_files containing data files for common 'd' and
+   * 'b' partitions.
+   *
+   * <p>If not set, manifests will be rewritten in the order of the transforms in the table's
+   * current partition spec.
+   *
+   * @param partitionFieldClustering Exact transformed column names used for partitioning; not the
+   *     raw column names that partitions are derived from. E.G. supply 'data_bucket' and not 'data'
+   *     for a bucket(N, data) partition * definition
+   * @return this for method chaining
+   */
+  default RewriteManifests clusterBy(List<String> partitionFieldClustering) {
+    throw new UnsupportedOperationException(
+        this.getClass().getName() + " doesn't implement clusterBy(List<String>)");
+  }
+
+  /**
+   * Rewrite manifests in a given order, dictated by a custom Function
+   *
+   * <p>Supply a Function which will apply its own custom clustering logic based on supplied {@link
+   * org.apache.iceberg.DataFile} attributes.
+   *
+   * @param clusterStrategyFunction A Function that returns a String to be used for manifest
+   *     clustering
+   * @return this method for chaining
+   */
+  default RewriteManifests clusterBy(Function<DataFile, String> clusterStrategyFunction) {
```

Comment:
I don't feel good about opening up a public api like this that takes an arbitrary function unless we have a very good reason to. Do we need this kind of flexability?

Reply:
I can say that there are reasons to, but it is up to you to decide if they are very good reasons. 

For example, I know that a primary use case for `bucket(someId, 10000)` partitioned tables will require reading `bucket_someId=1` and `bucket_someId=999` data together, so I'll want to cluster my planning around this use case. 

Potentially simpler, say I'm partitioning by `month(timestamp)`. I can take the min/max values of a data file and cluster the files in a given month partition by day for more efficient query planning. This option lets power users take what Iceberg gives you to a new level.

Reply:
I should also mention that this is a feature in the public Java API, it just doesn't scale for large tables which is why I wanted to implement a Spark version 

https://github.com/apache/iceberg/blob/5e279c868f8e2087f88ae779d8cc8474768bc5e2/api/src/main/java/org/apache/iceberg/RewriteManifests.java#L49

Reply:
@zachdisc I really don't want to widen the api if we can at all help it, adding a Function<> api is something that will be hard for us to walk back. I would feel more comfortable if we started with partition_transforms and consider opening things up more as community demand becomes a bit more pronounced. 

I think a lot of your use cases will work better if we allow users to sort on any column. We can't really do that at the moment, but we have a lot of plans for V4 metadata that will allow us to bring all column min/maxes up to the manifest_list level.

Reply:
Thanks @RussellSpitzer. Since it is just you and I here I'll happily disagree and commit. I'll get rid of arbitrary function based reordering.